File: Leet_1512/solution.py
class Solution:
    def numIdenticalPairs(self, nums: list[int]) -> int:
        """
        Counts the number of good pairs in the given list of integers.
        A pair (i, j) is called good if nums[i] == nums[j] and i < j.
        Args:
            nums (list[int]): A list of integers.
        Returns:
            int: The number of good pairs in the list.
        """
        count: int = 0 
        table: dict[int, int] = {}

        for num in nums:
            if num not in table:
                table[num] = 1
            else:
                count += table[num]
                table[num] += 1

        return count


def main() -> None:
    print("1512. Number of Good Pairs")

    sol = Solution()
    nums = [1,2,3,1,1,3]
    print(sol.numIdenticalPairs(nums))


if __name__ == "__main__":
    main()
        

File: Leet_1512/solution.c
/*
1512. Number of Good Pairs

Given an array of integers nums, return the number of good pairs.
A pair (i, j) is called good if nums[i] == nums[j] and i < j.

Constraints:
    1 <= nums.length <= 100
    1 <= nums[i] <= 100
*/


int numIdenticalPairs(int* nums, int numsSize) 
{
    /*get max number */
    int maxNumber = -1;
    for(int i = 0; i < numsSize; i++)
    {
        if(nums[i] > maxNumber)
        {
            maxNumber = nums[i];
        }
    }

    /*update frequency table*/
    int count = 0;
    int* table = calloc(maxNumber + 1, sizeof(int));

    for(int i = 0; i < numsSize; i++)
    {
        if(table[nums[i]] == 0)
        {
            table[nums[i]] = 1;
        }
        else
        {
            count += table[nums[i]];
            table[nums[i]]++;
        }
    }

    free(table);
    
    return count; 
}